diff --git a/Lib/ctypes/__init__.py b/Lib/ctypes/__init__.py
index e24cfd2..f234797 100644
--- a/Lib/ctypes/__init__.py
+++ b/Lib/ctypes/__init__.py
@@ -326,6 +326,29 @@ def ARRAY(typ, len):
 ################################################################
 
 
+def _dlopen_from_zip(name, mode):
+    import tempfile
+    import shutil
+    # TODO: linux only
+    with open(name, 'rb') as srcfp:
+        fd, tmpname = tempfile.mkstemp('.so')
+        try:
+            with _os.fdopen(fd, 'wb') as dstfp:
+                shutil.copyfileobj(srcfp, dstfp)
+            return _dlopen(tmpname, mode)
+        finally:
+            try:
+                _os.unlink(tmpname)
+            except OSError:
+                pass
+
+def dlopen_from_zip(name, mode):
+    try:
+        return _dlopen_from_zip(name, mode)
+    except (OSError, IOError):
+        return
+
+
 class CDLL(object):
     """An instance of this class represents a loaded dll/shared
     library, exporting functions using the standard C calling
@@ -359,7 +382,13 @@ class CDLL(object):
         self._FuncPtr = _FuncPtr
 
         if handle is None:
-            self._handle = _dlopen(self._name, mode)
+            try:
+                self._handle = _dlopen(self._name, mode)
+            except OSError:
+                handle = dlopen_from_zip(self._name, mode)
+                if handle is None:
+                    raise
+                self._handle = handle
         else:
             self._handle = handle
 
diff --git a/Lib/inspect.py b/Lib/inspect.py
index b08e145..271eff9 100644
--- a/Lib/inspect.py
+++ b/Lib/inspect.py
@@ -456,6 +456,9 @@ def getsourcefile(object):
     # or it is in the linecache
     if filename in linecache.cache:
         return filename
+    # a workaround for packages like django debug toolbar
+    if filename.startswith('<pyrun>'):
+        return filename
 
 def getabsfile(object, _filename=None):
     """Return an absolute path to the source or compiled file for an object.
diff --git a/Lib/subprocess.py b/Lib/subprocess.py
index 187566b..cd707c9 100644
--- a/Lib/subprocess.py
+++ b/Lib/subprocess.py
@@ -1195,6 +1195,21 @@ class Popen(object):
                     except:
                         pass
 
+        def __get_executable_from_zip(self, name):
+            import tempfile
+            import shutil
+            with open(name, 'rb') as srcfp:
+                fd, tmpname = tempfile.mkstemp()
+                with os.fdopen(fd, 'wb') as dstfp:
+                    shutil.copyfileobj(srcfp, dstfp)
+                os.chmod(tmpname, 0o700)
+                return tmpname
+
+        def _get_executable_from_zip(self, name):
+            try:
+                return self.__get_executable_from_zip(name)
+            except (OSError, IOError):
+                return
 
         def _execute_child(self, args, executable, preexec_fn, close_fds,
                            cwd, env, universal_newlines,
@@ -1202,6 +1217,63 @@ class Popen(object):
                            p2cread, p2cwrite,
                            c2pread, c2pwrite,
                            errread, errwrite):
+            # TODO: massive hack here: if cwd points inside zip, clear
+            # it. may be a very unexpected behaviour
+            if cwd is not None and cwd.startswith(sys.executable):
+                cwd = None
+            # TODO: another shameless hack. most existing code assumes
+            # sys.executable is an usable python binary that can be
+            # called like "sys.executable -m module", etc. we support
+            # it by adding EXXO_FORCE_STANDALONE to env
+            if executable == sys.executable or args[0] == sys.executable:
+                if env is None:
+                    env = {}
+                # it's possible to disable above behaviour in a dirty
+                # way: by settings EXXO_FORCE_STANDALONE to an empty
+                # string
+                if 'EXXO_FORCE_STANDALONE' not in env:
+                    env['EXXO_FORCE_STANDALONE'] = '1'
+            try:
+                self.__execute_child(args, executable, preexec_fn, close_fds,
+                                     cwd, env, universal_newlines,
+                                     startupinfo, creationflags, shell, to_close,
+                                     p2cread, p2cwrite,
+                                     c2pread, c2pwrite,
+                                     errread, errwrite)
+            except OSError as e:
+                if e.errno != os.errno.ENOTDIR:
+                    raise
+                exe = executable
+                if exe is None:
+                    if isinstance(args, basestring):
+                        raise
+                    exe = args[0]
+                exe = self._get_executable_from_zip(exe)
+                if exe is None:
+                    raise
+                if executable is not None:
+                    executable = exe
+                args = list(args)
+                args[0] = exe
+                try:
+                    self.__execute_child(args, executable, preexec_fn, close_fds,
+                                         cwd, env, universal_newlines,
+                                         startupinfo, creationflags, shell, to_close,
+                                         p2cread, p2cwrite,
+                                         c2pread, c2pwrite,
+                                         errread, errwrite)
+                finally:
+                    try:
+                        os.unlink(exe)
+                    except OSError:
+                        pass
+
+        def __execute_child(self, args, executable, preexec_fn, close_fds,
+                            cwd, env, universal_newlines,
+                            startupinfo, creationflags, shell, to_close,
+                            p2cread, p2cwrite,
+                            c2pread, c2pwrite,
+                            errread, errwrite):
             """Execute program (POSIX version)"""
 
             if isinstance(args, types.StringTypes):
diff --git a/Modules/_ssl.c b/Modules/_ssl.c
index f291352..0918134 100644
--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -2713,6 +2713,38 @@ _add_ca_certs(PySSLContext *self, void *data, Py_ssize_t len,
 
 
 static PyObject *
+get_cert_from_zip(PyObject *cafile)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "get_file");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", cafile);
+    Py_DECREF(func);
+    if (obj == NULL) {
+        PyErr_Clear();
+        return NULL;
+    }
+
+    func = PyObject_GetAttrString(obj, "read");
+    Py_DECREF(obj);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, NULL);
+    Py_DECREF(func);
+    return obj;
+}
+
+
+static PyObject *
 load_verify_locations(PySSLContext *self, PyObject *args, PyObject *kwds)
 {
     char *kwlist[] = {"cafile", "capath", "cadata", NULL};
@@ -2741,18 +2773,34 @@ load_verify_locations(PySSLContext *self, PyObject *args, PyObject *kwds)
     }
 
     if (cafile) {
-        if (PyString_Check(cafile)) {
-            Py_INCREF(cafile);
-            cafile_bytes = cafile;
-        } else {
-            PyObject *u = PyUnicode_FromObject(cafile);
-            if (!u)
-                goto error;
-            cafile_bytes = PyUnicode_AsEncodedString(
-                u, Py_FileSystemDefaultEncoding, NULL);
-            Py_DECREF(u);
-            if (!cafile_bytes)
+        PyObject *cazipdata;
+
+        cazipdata = get_cert_from_zip(cafile);
+        if (cazipdata) {
+            char *cadata_buf;
+            Py_ssize_t cadata_len;
+
+            if (PyBytes_AsStringAndSize(cazipdata, &cadata_buf, &cadata_len) < 0 ||
+                _add_ca_certs(self, cadata_buf, cadata_len, SSL_FILETYPE_PEM) < 0) {
+                Py_DECREF(cazipdata);
                 goto error;
+            }
+
+            Py_DECREF(cazipdata);
+        } else {
+            if (PyString_Check(cafile)) {
+                Py_INCREF(cafile);
+                cafile_bytes = cafile;
+            } else {
+                PyObject *u = PyUnicode_FromObject(cafile);
+                if (!u)
+                    goto error;
+                cafile_bytes = PyUnicode_AsEncodedString(
+                    u, Py_FileSystemDefaultEncoding, NULL);
+                Py_DECREF(u);
+                if (!cafile_bytes)
+                    goto error;
+            }
         }
     }
     if (capath) {
diff --git a/Modules/makesetup b/Modules/makesetup
index 8862c36..2afec2d 100755
--- a/Modules/makesetup
+++ b/Modules/makesetup
@@ -176,6 +176,7 @@ sed -e 's/[ 	]*#.*//' -e '/^[ 	]*$/d' |
 			*.c++)		srcs="$srcs $arg";;
 			*.cxx)		srcs="$srcs $arg";;
 			*.cpp)		srcs="$srcs $arg";;
+			*.S)		srcs="$srcs $arg";;
 			\$*)		libs="$libs $arg"
 					cpps="$cpps $arg";;
 			*.*)		echo 1>&2 "bad word $arg in $line"
@@ -206,6 +207,7 @@ sed -e 's/[ 	]*#.*//' -e '/^[ 	]*$/d' |
 			*.cxx) obj=`basename $src .cxx`.o; cc='$(CXX)';;
 			*.cpp) obj=`basename $src .cpp`.o; cc='$(CXX)';;
 			*.m)   obj=`basename $src .m`.o; cc='$(CC)';; # Obj-C
+			*.S)   obj=`basename $src .S`.o; cc='$(CC)';;
 			*)     continue;;
 			esac
 			obj="$srcdir/$obj"
diff --git a/Modules/posixmodule.c b/Modules/posixmodule.c
index 4fc3ef7..4bce72e 100644
--- a/Modules/posixmodule.c
+++ b/Modules/posixmodule.c
@@ -1629,6 +1629,25 @@ IsUNCRootW(Py_UNICODE *path, int pathlen)
 #endif /* MS_WINDOWS */
 
 static PyObject *
+stat_inzip(PyObject *filename)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "stat_file");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", filename);
+    Py_DECREF(func);
+    return obj;
+}
+
+static PyObject *
 posix_do_stat(PyObject *self, PyObject *args,
               char *format,
 #ifdef __VMS
@@ -1678,7 +1697,13 @@ posix_do_stat(PyObject *self, PyObject *args,
 #ifdef MS_WINDOWS
         result = win32_error("stat", pathfree);
 #else
-        result = posix_error_with_filename(pathfree);
+        if (errno == ENOTDIR) {
+            PyObject *fn;
+            fn = PyString_FromString(pathfree);
+            result = stat_inzip(fn);
+            Py_DECREF(fn);
+        } else
+            result = posix_error_with_filename(pathfree);
 #endif
     }
     else
@@ -2276,6 +2301,25 @@ posix_link(PyObject *self, PyObject *args)
 #endif /* HAVE_LINK */
 
 
+static PyObject *
+listdir_inzip(PyObject *dir)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "listdir");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", dir);
+    Py_DECREF(func);
+    return obj;
+}
+
 PyDoc_STRVAR(posix_listdir__doc__,
 "listdir(path) -> list_of_strings\n\n\
 Return a list containing the names of the entries in the directory.\n\
@@ -2532,7 +2576,15 @@ posix_listdir(PyObject *self, PyObject *args)
     dirp = opendir(name);
     Py_END_ALLOW_THREADS
     if (dirp == NULL) {
-        return posix_error_with_allocated_filename(name);
+        if (errno == ENOTDIR) {
+            PyObject *res, *pyname;
+            pyname = PyString_FromString(name);
+            res = listdir_inzip(pyname);
+            Py_DECREF(pyname);
+            PyMem_Free(name);
+            return res;
+        } else
+            return posix_error_with_allocated_filename(name);
     }
     if ((d = PyList_New(0)) == NULL) {
         Py_BEGIN_ALLOW_THREADS
diff --git a/Python/bltinmodule.c b/Python/bltinmodule.c
index 2977bcd..44eb09f 100644
--- a/Python/bltinmodule.c
+++ b/Python/bltinmodule.c
@@ -1500,9 +1500,38 @@ Return the octal representation of an integer or long integer.");
 
 
 static PyObject *
+get_inzip_file(PyObject *filename)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "get_file");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", filename);
+    Py_DECREF(func);
+    return obj;
+}
+
+
+static PyObject *
 builtin_open(PyObject *self, PyObject *args, PyObject *kwds)
 {
-    return PyObject_Call((PyObject*)&PyFile_Type, args, kwds);
+    PyObject *file_object;
+
+    file_object = PyObject_Call((PyObject*)&PyFile_Type, args, kwds);
+
+    if (PyErr_Occurred() && PyErr_ExceptionMatches(PyExc_IOError) && errno == ENOTDIR) {
+        PyErr_Clear();
+        file_object = get_inzip_file(PyTuple_GET_ITEM(args, 0));
+    }
+
+    return file_object;
 }
 
 PyDoc_STRVAR(open_doc,
diff --git a/Python/import.c b/Python/import.c
index 96f7d47..2ce34aa 100644
--- a/Python/import.c
+++ b/Python/import.c
@@ -1310,6 +1310,24 @@ static int case_ok(char *, Py_ssize_t, Py_ssize_t, char *);
 static int find_init_module(char *); /* Forward */
 static struct filedescr importhookdescr = {"", "", IMP_HOOK};
 
+static PyObject *_get_zipimporter(void)
+{
+    static PyObject *zipimporter = NULL;
+
+    if (zipimporter == NULL) {
+        PyObject *mod;
+
+        mod = PyImport_ImportModule("zipimport");
+        if (mod == NULL)
+            return NULL;
+
+        zipimporter = PyObject_GetAttrString(mod, "zipimporter");
+        Py_DECREF(mod);
+    }
+
+    return zipimporter;
+}
+
 static struct filedescr *
 find_module(char *fullname, char *subname, PyObject *path, char *buf,
             size_t buflen, FILE **p_fp, PyObject **p_loader)
@@ -1323,6 +1341,7 @@ find_module(char *fullname, char *subname, PyObject *path, char *buf,
     static struct filedescr fd_frozen = {"", "", PY_FROZEN};
     static struct filedescr fd_builtin = {"", "", C_BUILTIN};
     static struct filedescr fd_package = {"", "", PKG_DIRECTORY};
+    static struct filedescr fd_zip = {"", "", ZIP_IMPORT};
     char *name;
 #if defined(PYOS_OS2)
     size_t saved_len;
@@ -1486,18 +1505,25 @@ find_module(char *fullname, char *subname, PyObject *path, char *buf,
             }
             /* Note: importer is a borrowed reference */
             if (importer != Py_None) {
-                PyObject *loader;
+                PyObject *loader, *zipimporter;
                 loader = PyObject_CallMethod(importer,
                                              "find_module",
                                              "s", fullname);
                 Py_XDECREF(copy);
                 if (loader == NULL)
                     goto error_exit;  /* error */
+                zipimporter = _get_zipimporter();
+                if (zipimporter == NULL)
+                    goto error_exit;
                 if (loader != Py_None) {
                     /* a loader was found */
                     *p_loader = loader;
                     PyMem_FREE(name);
-                    return &importhookdescr;
+                    /* special case zipimporter */
+                    if (PyObject_IsInstance(loader, zipimporter))
+                        return &fd_zip;
+                    else
+                        return &importhookdescr;
                 }
                 Py_DECREF(loader);
                 continue;
@@ -1974,7 +2000,8 @@ load_module(char *name, FILE *fp, char *pathname, int type, PyObject *loader)
         Py_INCREF(m);
         break;
 
-    case IMP_HOOK: {
+    case IMP_HOOK:
+    case ZIP_IMPORT: {
         if (loader == NULL) {
             PyErr_SetString(PyExc_ImportError,
                             "import hook without loader");
@@ -2964,7 +2991,7 @@ static PyObject *
 call_find_module(char *name, PyObject *path)
 {
     extern int fclose(FILE *);
-    PyObject *fob, *ret;
+    PyObject *fob, *ret, *loader;
     struct filedescr *fdp;
     char *pathname;
     FILE *fp = NULL;
@@ -2976,11 +3003,12 @@ call_find_module(char *name, PyObject *path)
     pathname[0] = '\0';
     if (path == Py_None)
         path = NULL;
-    fdp = find_module(NULL, name, path, pathname, MAXPATHLEN+1, &fp, NULL);
+    fdp = find_module(name, name, path, pathname, MAXPATHLEN+1, &fp, &loader);
     if (fdp == NULL) {
         PyMem_FREE(pathname);
         return NULL;
     }
+    Py_XDECREF(loader);
     if (fp != NULL) {
         fob = PyFile_FromFile(fp, pathname, fdp->mode, fclose);
         if (fob == NULL) {
@@ -2992,8 +3020,24 @@ call_find_module(char *name, PyObject *path)
         fob = Py_None;
         Py_INCREF(fob);
     }
+    if (fdp->type == ZIP_IMPORT) {
+        char *newpath;
+        size_t bufsize;
+
+        bufsize = strlen(pathname) + strlen(name) + 2;
+        newpath = PyMem_MALLOC(bufsize);
+        if (newpath == NULL) {
+            PyErr_NoMemory();
+            goto exit;
+        }
+
+        PyOS_snprintf(newpath, bufsize, "%s%c%s", pathname, SEP, name);
+        PyMem_FREE(pathname);
+        pathname = newpath;
+    }
     ret = Py_BuildValue("Os(ssi)",
                   fob, pathname, fdp->suffix, fdp->mode, fdp->type);
+exit:
     Py_DECREF(fob);
     PyMem_FREE(pathname);
     return ret;
diff --git a/Python/importdl.h b/Python/importdl.h
index b4d21be..07f4468 100644
--- a/Python/importdl.h
+++ b/Python/importdl.h
@@ -17,7 +17,8 @@ enum filetype {
     C_BUILTIN,
     PY_FROZEN,
     PY_CODERESOURCE, /* Mac only */
-    IMP_HOOK
+    IMP_HOOK,
+    ZIP_IMPORT
 };
 
 struct filedescr {
