diff --git a/Lib/_exxo_hack.py b/Lib/_exxo_hack.py
new file mode 100644
index 0000000..b8f294a
--- /dev/null
+++ b/Lib/_exxo_hack.py
@@ -0,0 +1,63 @@
+import sys
+import os
+import posix
+import stat
+import zipfile
+import errno
+
+
+if zipfile.is_zipfile(sys.executable):
+    exe_zip = zipfile.ZipFile(sys.executable, 'r')
+    exe_names = exe_zip.namelist()
+zip_cache = {}
+
+
+def cached(func):
+    cache = zip_cache.setdefault(func, {})
+    def wrapped(*args):
+        ent = cache.get(args)
+        if ent is None:
+            ent = func(*args)
+            cache[args] = ent
+        return ent
+    return wrapped
+
+
+@cached
+def _get_inzip_path(filename, exc):
+    if not zipfile.is_zipfile(sys.executable):
+        raise exc(errno.ENOENT, filename)
+    inzip_path = filename[len(sys.executable):].lstrip('/')
+    if inzip_path not in exe_names:
+        # try a directory
+        inzip_path = inzip_path + '/'
+        if inzip_path not in exe_names:
+            raise exc(errno.ENOENT, filename)
+    is_dir = inzip_path.endswith('/')
+    return inzip_path, is_dir
+
+
+def get_file(filename):
+    inzip_path, _ = _get_inzip_path(filename, IOError)
+    return exe_zip.open(inzip_path)
+
+
+@cached
+def stat_file(filename):
+    inzip_path, is_dir = _get_inzip_path(filename, OSError)
+    info = exe_zip.getinfo(inzip_path)
+    stat_result = list(os.stat(sys.executable))
+    stat_result[6] = info.file_size
+    if is_dir:
+        stat_result[0] &= ~(stat.S_IFREG | stat.S_IFLNK)
+        stat_result[0] |= stat.S_IFDIR
+    return posix.stat_result(stat_result)
+
+
+@cached
+def listdir(directory):
+    fixed_directory = directory.rstrip('/') + '/'
+    inzip_path, is_dir = _get_inzip_path(fixed_directory, OSError)
+    if not is_dir:
+        raise OSError(errno.ENOTDIR, directory)
+    return [n[len(inzip_path):] for n in exe_names if n != inzip_path and n.startswith(inzip_path)]
diff --git a/Lib/ctypes/__init__.py b/Lib/ctypes/__init__.py
index 0d86078..2b55fe9 100644
--- a/Lib/ctypes/__init__.py
+++ b/Lib/ctypes/__init__.py
@@ -311,6 +311,29 @@ def ARRAY(typ, len):
 ################################################################
 
 
+def _dlopen_from_zip(name, mode):
+    import tempfile
+    import shutil
+    # TODO: linux only
+    with open(name, 'rb') as srcfp:
+        fd, tmpname = tempfile.mkstemp('.so')
+        try:
+            with _os.fdopen(fd, 'wb') as dstfp:
+                shutil.copyfileobj(srcfp, dstfp)
+            return _dlopen(tmpname, mode)
+        finally:
+            try:
+                _os.unlink(tmpname)
+            except OSError:
+                pass
+
+def dlopen_from_zip(name, mode):
+    try:
+        return _dlopen_from_zip(name, mode)
+    except (OSError, IOError):
+        return
+
+
 class CDLL(object):
     """An instance of this class represents a loaded dll/shared
     library, exporting functions using the standard C calling
@@ -344,7 +367,13 @@ class CDLL(object):
         self._FuncPtr = _FuncPtr
 
         if handle is None:
-            self._handle = _dlopen(self._name, mode)
+            try:
+                self._handle = _dlopen(self._name, mode)
+            except OSError:
+                handle = dlopen_from_zip(self._name, mode)
+                if handle is None:
+                    raise
+                self._handle = handle
         else:
             self._handle = handle
 
diff --git a/Lib/imp.py b/Lib/imp.py
index e264391..8b1568a 100644
--- a/Lib/imp.py
+++ b/Lib/imp.py
@@ -27,6 +27,7 @@ import sys
 import tokenize
 import types
 import warnings
+from zipimport import zipimporter
 
 warnings.warn("the imp module is deprecated in favour of importlib; "
               "see the module's documentation for alternative uses",
@@ -278,6 +279,17 @@ def find_module(name, path=None):
             path = sys.path
 
     for entry in path:
+        for meta in sys.path_hooks:
+            if meta is not zipimporter:
+                continue
+            try:
+                loader = meta(entry)
+            except ImportError:
+                continue
+            if loader.find_module(name) is None:
+                continue
+            newpath = os.path.join(entry, name)
+            return None, newpath, ('', '', IMP_HOOK)
         package_directory = os.path.join(entry, name)
         for suffix in ['.py', machinery.BYTECODE_SUFFIXES[0]]:
             package_file_name = '__init__' + suffix
diff --git a/Lib/inspect.py b/Lib/inspect.py
index e830eb6..d36e7b7 100644
--- a/Lib/inspect.py
+++ b/Lib/inspect.py
@@ -674,6 +674,9 @@ def getsourcefile(object):
     # or it is in the linecache
     if filename in linecache.cache:
         return filename
+    # a workaround for packages like django debug toolbar
+    if filename.startswith('<pyrun>'):
+        return filename
 
 def getabsfile(object, _filename=None):
     """Return an absolute path to the source or compiled file for an object.
diff --git a/Lib/subprocess.py b/Lib/subprocess.py
index d8d6ab2..2e63485 100644
--- a/Lib/subprocess.py
+++ b/Lib/subprocess.py
@@ -1424,6 +1424,21 @@ class Popen(object):
                     c2pread, c2pwrite,
                     errread, errwrite)
 
+        def __get_executable_from_zip(self, name):
+            import tempfile
+            import shutil
+            with open(name, 'rb') as srcfp:
+                fd, tmpname = tempfile.mkstemp()
+                with os.fdopen(fd, 'wb') as dstfp:
+                    shutil.copyfileobj(srcfp, dstfp)
+                os.chmod(tmpname, 0o700)
+                return tmpname
+
+        def _get_executable_from_zip(self, name):
+            try:
+                return self.__get_executable_from_zip(name)
+            except (OSError, IOError):
+                return
 
         def _execute_child(self, args, executable, preexec_fn, close_fds,
                            pass_fds, cwd, env,
@@ -1432,6 +1447,66 @@ class Popen(object):
                            c2pread, c2pwrite,
                            errread, errwrite,
                            restore_signals, start_new_session):
+            # TODO: massive hack here: if cwd points inside zip, clear
+            # it. may be a very unexpected behaviour
+            if cwd is not None and cwd.startswith(sys.executable):
+                cwd = None
+            # TODO: another shameless hack. most existing code assumes
+            # sys.executable is an usable python binary that can be
+            # called like "sys.executable -m module", etc. we support
+            # it by adding EXXO_FORCE_STANDALONE to env
+            if executable == sys.executable or args[0] == sys.executable:
+                if env is None:
+                    env = {}
+                # it's possible to disable above behaviour in a dirty
+                # way: by settings EXXO_FORCE_STANDALONE to an empty
+                # string
+                if 'EXXO_FORCE_STANDALONE' not in env:
+                    env['EXXO_FORCE_STANDALONE'] = '1'
+            try:
+                self.__execute_child(args, executable, preexec_fn, close_fds,
+                                     pass_fds, cwd, env,
+                                     startupinfo, creationflags, shell,
+                                     p2cread, p2cwrite,
+                                     c2pread, c2pwrite,
+                                     errread, errwrite,
+                                     restore_signals, start_new_session)
+            except OSError as e:
+                if e.errno != os.errno.ENOTDIR:
+                    raise
+                exe = executable
+                if exe is None:
+                    if isinstance(args, str):
+                        raise
+                    exe = args[0]
+                exe = self._get_executable_from_zip(exe)
+                if exe is None:
+                    raise
+                if executable is not None:
+                    executable = exe
+                args = list(args)
+                args[0] = exe
+                try:
+                    self.__execute_child(args, executable, preexec_fn, close_fds,
+                                         pass_fds, cwd, env,
+                                         startupinfo, creationflags, shell,
+                                         p2cread, p2cwrite,
+                                         c2pread, c2pwrite,
+                                         errread, errwrite,
+                                         restore_signals, start_new_session)
+                finally:
+                    try:
+                        os.unlink(exe)
+                    except OSError:
+                        pass
+
+        def __execute_child(self, args, executable, preexec_fn, close_fds,
+                            pass_fds, cwd, env,
+                            startupinfo, creationflags, shell,
+                            p2cread, p2cwrite,
+                            c2pread, c2pwrite,
+                            errread, errwrite,
+                            restore_signals, start_new_session):
             """Execute program (POSIX version)"""
 
             if isinstance(args, (str, bytes)):
diff --git a/Makefile.pre.in b/Makefile.pre.in
index f0ddddb..c717461 100644
--- a/Makefile.pre.in
+++ b/Makefile.pre.in
@@ -1128,9 +1128,9 @@ bininstall: altbininstall
 	(cd $(DESTDIR)$(BINDIR); $(LN) -s python$(VERSION)$(EXE) python3$(EXE))
 	-if test "$(VERSION)" != "$(LDVERSION)"; then \
 		rm -f $(DESTDIR)$(BINDIR)/python$(VERSION)-config; \
-		(cd $(DESTDIR)$(BINDIR); $(LN) -s python$(LDVERSION)-config python$(VERSION)-config); \
+		(cd $(DESTDIR)$(BINDIR); $(LN) -sf python$(LDVERSION)-config python$(VERSION)-config); \
 		rm -f $(DESTDIR)$(LIBPC)/python-$(LDVERSION).pc; \
-		(cd $(DESTDIR)$(LIBPC); $(LN) -s python-$(VERSION).pc python-$(LDVERSION).pc); \
+		(cd $(DESTDIR)$(LIBPC); $(LN) -sf python-$(VERSION).pc python-$(LDVERSION).pc); \
 	fi
 	-rm -f $(DESTDIR)$(BINDIR)/python3-config
 	(cd $(DESTDIR)$(BINDIR); $(LN) -s python$(VERSION)-config python3-config)
diff --git a/Modules/_io/_iomodule.c b/Modules/_io/_iomodule.c
index 65c955a..34ff398 100644
--- a/Modules/_io/_iomodule.c
+++ b/Modules/_io/_iomodule.c
@@ -52,6 +52,27 @@ PyObject *_PyIO_empty_str;
 PyObject *_PyIO_empty_bytes;
 PyObject *_PyIO_zero;
 
+
+static PyObject *
+get_inzip_file(PyObject *filename)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "get_file");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", filename);
+    Py_DECREF(func);
+    return obj;
+}
+
+
 
 PyDoc_STRVAR(module_doc,
 "The io module provides the Python interfaces to stream handling. The\n"
@@ -351,6 +372,11 @@ _io_open_impl(PyObject *module, PyObject *file, const char *mode,
     /* Create the Raw file stream */
     raw = PyObject_CallFunction((PyObject *)&PyFileIO_Type,
                                 "OsiO", file, rawmode, closefd, opener);
+    if (PyErr_Occurred() && PyErr_ExceptionMatches(PyExc_OSError) && errno == ENOTDIR) {
+        PyErr_Clear();
+        raw = get_inzip_file(file);
+        buffering = 65536;
+    }
     if (raw == NULL)
         return NULL;
     result = raw;
diff --git a/Modules/_ssl.c b/Modules/_ssl.c
index 391034e..b15ddee 100644
--- a/Modules/_ssl.c
+++ b/Modules/_ssl.c
@@ -2963,6 +2963,38 @@ _add_ca_certs(PySSLContext *self, void *data, Py_ssize_t len,
 }
 
 
+static PyObject *
+get_cert_from_zip(PyObject *cafile)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "get_file");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", cafile);
+    Py_DECREF(func);
+    if (obj == NULL) {
+        PyErr_Clear();
+        return NULL;
+    }
+
+    func = PyObject_GetAttrString(obj, "read");
+    Py_DECREF(obj);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, NULL);
+    Py_DECREF(func);
+    return obj;
+}
+
+
 /*[clinic input]
 _ssl._SSLContext.load_verify_locations
     cafile: object = NULL
@@ -3044,6 +3076,25 @@ _ssl__SSLContext_load_verify_locations_impl(PySSLContext *self,
         }
     }
 
+    if (cafile) {
+        PyObject *cazipdata;
+
+        cazipdata = get_cert_from_zip(cafile);
+        if (cazipdata) {
+            char *cadata_buf;
+            Py_ssize_t cadata_len;
+
+            if (PyBytes_AsStringAndSize(cazipdata, &cadata_buf, &cadata_len) < 0 ||
+                _add_ca_certs(self, cadata_buf, cadata_len, SSL_FILETYPE_PEM) < 0) {
+                Py_DECREF(cazipdata);
+                goto error;
+            }
+
+            Py_DECREF(cazipdata);
+            cafile = NULL;
+        }
+    }
+
     /* load cafile or capath */
     if (cafile || capath) {
         if (cafile)
diff --git a/Modules/makesetup b/Modules/makesetup
index 90db42e..3c93828 100755
--- a/Modules/makesetup
+++ b/Modules/makesetup
@@ -176,6 +176,7 @@ sed -e 's/[ 	]*#.*//' -e '/^[ 	]*$/d' |
 			*.c++)		srcs="$srcs $arg";;
 			*.cxx)		srcs="$srcs $arg";;
 			*.cpp)		srcs="$srcs $arg";;
+			*.S)		srcs="$srcs $arg";;
 			\$*)		libs="$libs $arg"
 					cpps="$cpps $arg";;
 			*.*)		echo 1>&2 "bad word $arg in $line"
@@ -206,6 +207,7 @@ sed -e 's/[ 	]*#.*//' -e '/^[ 	]*$/d' |
 			*.cxx) obj=`basename $src .cxx`.o; cc='$(CXX)';;
 			*.cpp) obj=`basename $src .cpp`.o; cc='$(CXX)';;
 			*.m)   obj=`basename $src .m`.o; cc='$(CC)';; # Obj-C
+			*.S)   obj=`basename $src .S`.o; cc='$(CC)';;
 			*)     continue;;
 			esac
 			obj="$srcdir/$obj"
diff --git a/Modules/posixmodule.c b/Modules/posixmodule.c
index b854caf..e016b6f 100644
--- a/Modules/posixmodule.c
+++ b/Modules/posixmodule.c
@@ -2100,6 +2100,24 @@ _pystat_fromstructstat(STRUCT_STAT *st)
 
 /* POSIX methods */
 
+static PyObject *
+stat_inzip(PyObject *filename)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "stat_file");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", filename);
+    Py_DECREF(func);
+    return obj;
+}
 
 static PyObject *
 posix_do_stat(char *function_name, path_t *path,
@@ -2146,7 +2164,10 @@ posix_do_stat(char *function_name, path_t *path,
     Py_END_ALLOW_THREADS
 
     if (result != 0) {
-        return path_error(path);
+        if (errno == ENOTDIR)
+            return stat_inzip(path->object);
+        else
+            return path_error(path);
     }
 
     return _pystat_fromstructstat(&st);
@@ -3452,6 +3473,26 @@ os_link_impl(PyObject *module, path_t *src, path_t *dst, int src_dir_fd,
 #endif
 
 
+static PyObject *
+listdir_inzip(PyObject *dir)
+{
+    PyObject *mod, *func, *obj;
+
+    mod = PyImport_ImportModule("_exxo_hack");
+    if (!mod)
+        return NULL;
+
+    func = PyObject_GetAttrString(mod, "listdir");
+    Py_DECREF(mod);
+    if (!func)
+        return NULL;
+
+    obj = PyObject_CallFunction(func, "O", dir);
+    Py_DECREF(func);
+    return obj;
+}
+
+
 #if defined(MS_WINDOWS) && !defined(HAVE_OPENDIR)
 static PyObject *
 _listdir_windows_no_opendir(path_t *path, PyObject *list)
@@ -3652,7 +3693,10 @@ _posix_listdir(path_t *path, PyObject *list)
     }
 
     if (dirp == NULL) {
-        list = path_error(path);
+        if (errno == ENOTDIR)
+            list = listdir_inzip(path->object);
+        else
+            list = path_error(path);
 #ifdef HAVE_FDOPENDIR
         if (fd != -1) {
             Py_BEGIN_ALLOW_THREADS
diff --git a/Modules/zipimport.c b/Modules/zipimport.c
index e840271..c7ef162 100644
--- a/Modules/zipimport.c
+++ b/Modules/zipimport.c
@@ -3,14 +3,16 @@
 #include "osdefs.h"
 #include "marshal.h"
 #include <time.h>
+#include <assert.h>
 
 
 #define IS_SOURCE   0x0
 #define IS_BYTECODE 0x1
 #define IS_PACKAGE  0x2
+#define IS_SOLIB    0x4
 
 struct st_zip_searchorder {
-    char suffix[14];
+    char suffix[64];
     int type;
 };
 
@@ -23,13 +25,19 @@ _Py_IDENTIFIER(replace);
    non-package .pyc, and .py entries. The .pyc entries
    are swapped by initzipimport() if we run in optimized mode. Also,
    '/' is replaced by SEP there. */
-static struct st_zip_searchorder zip_searchorder[] = {
-    {"/__init__.pyc", IS_PACKAGE | IS_BYTECODE},
-    {"/__init__.py", IS_PACKAGE | IS_SOURCE},
+static struct st_zip_searchorder *zip_searchorder;
+static struct st_zip_searchorder zip_searchorder_base[] = {
+    /* these two will be filled later in init_search_order() */
+    {"", IS_PACKAGE | IS_BYTECODE},
+    {"", IS_PACKAGE | IS_SOURCE},
     {".pyc", IS_BYTECODE},
-    {".py", IS_SOURCE},
-    {"", 0}
+    {".py", IS_SOURCE}
 };
+static const int zip_searchorder_base_len = sizeof(zip_searchorder_base) / sizeof(zip_searchorder_base[0]);
+
+extern PyObject *_PyImport_LoadDynamicModuleWithSpec(PyObject *spec, FILE *fp);
+extern const char *_PyImport_DynLoadFiletab[];
+extern const char * _PySys_ImplCacheTag;
 
 /* zipimporter object definition and support */
 
@@ -42,6 +50,9 @@ struct _zipimporter {
     PyObject *prefix;   /* file prefix: "a/sub/directory/",
                            encoded to the filesystem encoding */
     PyObject *files;    /* dict with file info {path: toc_entry} */
+
+    PyObject *make_spec;
+    PyObject *extension_loader;
 };
 
 static PyObject *ZipImportError;
@@ -53,6 +64,7 @@ static PyObject *read_directory(PyObject *archive);
 static PyObject *get_data(PyObject *archive, PyObject *toc_entry);
 static PyObject *get_module_code(ZipImporter *self, PyObject *fullname,
                                  int *p_ispackage, PyObject **p_modpath);
+static time_t get_mtime_of_source(ZipImporter *self, PyObject *path);
 
 
 #define ZipImporter_Check(op) PyObject_TypeCheck(op, &ZipImporter_Type)
@@ -67,6 +79,9 @@ zipimporter_init(ZipImporter *self, PyObject *args, PyObject *kwds)
 {
     PyObject *path, *files, *tmp;
     PyObject *filename = NULL;
+    PyObject *importlib = NULL;
+    PyObject *make_spec = NULL;
+    PyObject *extension_loader = NULL;
     Py_ssize_t len, flen;
 
     if (!_PyArg_NoKeywords("zipimporter()", kwds))
@@ -161,11 +176,33 @@ zipimporter_init(ZipImporter *self, PyObject *args, PyObject *kwds)
     else
         self->prefix = PyUnicode_New(0, 0);
     Py_DECREF(path);
+
+    importlib = PyImport_ImportModule("importlib.machinery");
+    if (!importlib)
+        return -1;
+
+    make_spec = PyObject_GetAttrString(importlib, "ModuleSpec");
+    if (!make_spec)
+        goto error_imp;
+
+    extension_loader = PyObject_GetAttrString(importlib, "ExtensionFileLoader");
+    if (!extension_loader)
+        goto error_imp;
+
+    Py_DECREF(importlib);
+
+    self->make_spec = make_spec;
+    self->extension_loader = extension_loader;
+
     return 0;
 
 error:
     Py_DECREF(path);
     Py_XDECREF(filename);
+error_imp:
+    Py_XDECREF(importlib);
+    Py_XDECREF(make_spec);
+    Py_XDECREF(extension_loader);
     return -1;
 }
 
@@ -286,20 +323,50 @@ check_is_directory(ZipImporter *self, PyObject* prefix, PyObject *path)
     return res;
 }
 
-/* Return some information about a module. */
+static PyObject *get_code_from_pycache(ZipImporter *self, PyObject *subname, time_t *mtime)
+{
+    PyObject *path, *fullpath, *item, *prefix;
+
+    prefix = PyUnicode_FromFormat("%U__pycache__%c", self->prefix, SEP);
+    if (prefix == NULL)
+        return NULL;
+
+    path = make_filename(prefix, subname);
+    Py_DECREF(prefix);
+    if (path == NULL)
+        return NULL;
+
+    fullpath = PyUnicode_FromFormat("%U.%s.pyc", path, _PySys_ImplCacheTag);
+    Py_DECREF(path);
+    if (fullpath == NULL)
+        return NULL;
+
+    item = PyDict_GetItem(self->files, fullpath);
+    if (mtime)
+        *mtime = get_mtime_of_source(self, fullpath);
+    Py_DECREF(fullpath);
+    if (item == NULL)
+        return NULL;
+
+    return item;
+}
+
 static enum zi_module_info
-get_module_info(ZipImporter *self, PyObject *fullname)
+get_module_info_from_pycache(ZipImporter *self, PyObject *subname)
+{
+    PyObject *item;
+
+    item = get_code_from_pycache(self, subname, NULL);
+    return item != NULL ? MI_MODULE : MI_NOT_FOUND;
+}
+
+static enum zi_module_info
+get_module_info_legacy(ZipImporter *self, PyObject *subname)
 {
-    PyObject *subname;
     PyObject *path, *fullpath, *item;
     struct st_zip_searchorder *zso;
 
-    subname = get_subname(fullname);
-    if (subname == NULL)
-        return MI_ERROR;
-
     path = make_filename(self->prefix, subname);
-    Py_DECREF(subname);
     if (path == NULL)
         return MI_ERROR;
 
@@ -309,6 +376,7 @@ get_module_info(ZipImporter *self, PyObject *fullname)
             Py_DECREF(path);
             return MI_ERROR;
         }
+
         item = PyDict_GetItem(self->files, fullpath);
         Py_DECREF(fullpath);
         if (item != NULL) {
@@ -319,10 +387,30 @@ get_module_info(ZipImporter *self, PyObject *fullname)
                 return MI_MODULE;
         }
     }
+
     Py_DECREF(path);
     return MI_NOT_FOUND;
 }
 
+/* Return some information about a module. */
+static enum zi_module_info
+get_module_info(ZipImporter *self, PyObject *fullname)
+{
+    PyObject *subname;
+    enum zi_module_info res;
+
+    subname = get_subname(fullname);
+    if (subname == NULL)
+        return MI_ERROR;
+
+    res = get_module_info_from_pycache(self, subname);
+    if (res == MI_NOT_FOUND)
+        res = get_module_info_legacy(self, subname);
+
+    Py_DECREF(subname);
+    return res;
+}
+
 typedef enum {
     FL_ERROR = -1,       /* error */
     FL_NOT_FOUND,        /* no loader or namespace portions found */
@@ -453,6 +541,128 @@ zipimporter_find_loader(PyObject *obj, PyObject *args)
     return result;
 }
 
+static PyObject *import_extension(ZipImporter *self, PyObject *fullname, PyObject *path)
+{
+    PyObject *spec = NULL;
+    PyObject *mod = NULL;
+    PyObject *loader = NULL;
+    PyObject *args = NULL;
+    PyObject *kw = NULL;
+
+    args = PyTuple_Pack(2, fullname, Py_None);
+    if (!args)
+        return NULL;
+
+    kw = PyDict_New();
+    if (!kw)
+        goto error;
+    if (PyDict_SetItemString(kw, "origin", path) < 0)
+        goto error;
+
+    /* spec = PyObject_CallFunction(self->make_spec, "OOO", fullname, Py_None, path); */
+    spec = PyObject_Call(self->make_spec, args, kw);
+    if (!spec)
+        goto error;
+
+    loader = PyObject_CallFunctionObjArgs(self->extension_loader, fullname, path, NULL);
+    if (!loader)
+        goto error;
+
+    mod = PyObject_CallMethod(loader, "create_module", "O", spec);
+
+error:
+    Py_XDECREF(args);
+    Py_XDECREF(kw);
+    Py_XDECREF(spec);
+    Py_XDECREF(loader);
+    return mod;
+}
+
+static PyObject *
+import_solib(ZipImporter *self, PyObject *fullname, PyObject *source)
+{
+    PyObject *mod = NULL;
+    PyObject *tmppath = NULL;
+    char tmpbuf[] = "/tmp/zipimportXXXXXX";
+    int tmp;
+    char *src;
+    Py_ssize_t srclen;
+
+    tmp = mkstemp(tmpbuf);
+    if (tmp < 0) {
+        PyErr_SetFromErrnoWithFilename(PyExc_OSError, tmpbuf);
+        return NULL;
+    }
+
+    if (PyBytes_AsStringAndSize(source, &src, &srclen) < 0)
+        goto exit;
+
+    while (srclen) {
+        int wrote = write(tmp, src, srclen);
+        if (wrote < 0) {
+            PyErr_SetFromErrnoWithFilename(PyExc_OSError, tmpbuf);
+            goto exit;
+        }
+        src += wrote;
+        srclen -= wrote;
+    }
+
+    tmppath = PyUnicode_FromString(tmpbuf);
+    if (!tmppath)
+        goto exit;
+
+    mod = import_extension(self, fullname, tmppath);
+
+exit:
+    close(tmp);
+    unlink(tmpbuf);
+    Py_XDECREF(tmppath);
+    return mod;
+}
+
+static PyObject *
+maybe_import_solib(ZipImporter *self, PyObject *fullname)
+{
+    PyObject *subname, *path, *fullpath;
+    PyObject *toc_entry, *data, *mod;
+    const char **suffix;
+
+    subname = get_subname(fullname);
+    if (subname == NULL)
+        Py_RETURN_NONE;
+
+    path = make_filename(self->prefix, subname);
+    Py_DECREF(subname);
+    if (path == NULL)
+        Py_RETURN_NONE;
+
+    for (suffix = _PyImport_DynLoadFiletab; *suffix != NULL; suffix++) {
+        fullpath = PyUnicode_FromFormat("%U%s", path, *suffix);
+        if (fullpath == NULL) {
+            Py_DECREF(path);
+            Py_RETURN_NONE;
+        }
+
+        toc_entry = PyDict_GetItem(self->files, fullpath);
+        Py_DECREF(fullpath);
+        if (toc_entry != NULL)
+            break;
+    }
+    Py_DECREF(path);
+    if (toc_entry == NULL)
+        return NULL;
+
+    data = get_data(self->archive, toc_entry);
+    if (data == NULL)
+        Py_RETURN_NONE;
+
+    mod = import_solib(self, fullname, data);
+    Py_DECREF(data);
+    if (!mod)
+        Py_RETURN_NONE;
+    return mod;
+}
+
 /* Load and return the module named by 'fullname'. */
 static PyObject *
 zipimporter_load_module(PyObject *obj, PyObject *args)
@@ -469,6 +679,14 @@ zipimporter_load_module(PyObject *obj, PyObject *args)
     if (PyUnicode_READY(fullname) == -1)
         return NULL;
 
+    mod = maybe_import_solib(self, fullname);
+    if (mod) {
+        if (mod == Py_None)
+            return NULL;
+        Py_DECREF(fullname);
+        return mod;
+    }
+
     code = get_module_code(self, fullname, &ispackage, &modpath);
     if (code == NULL)
         goto error;
@@ -1456,6 +1674,40 @@ get_code_from_data(ZipImporter *self, int ispackage, int isbytecode,
     return code;
 }
 
+/* Convert /path/to/__pycache/foo.pyc into /path/to/foo.pyc for
+ * compatibility reasons. Neither pkgutil nor pkg_resources will work
+ * with __pycache__ path at the moment. */
+static PyObject *
+strip_pycache_from_path(PyObject *path)
+{
+    wchar_t *s, *sub, *sub2;
+    PyObject *result = NULL, *newpath;
+
+    s = PyUnicode_AsWideCharString(path, NULL);
+    if (s == NULL)
+        return NULL;
+
+    sub = wcsstr(s, L"__pycache__");
+    if (sub == NULL) {
+        Py_INCREF(path);
+        result = path;
+        goto exit;
+    }
+
+    sub2 = sub + wcslen(L"__pycache__");
+    wmemmove(sub, sub2 + 1, wcslen(sub2) + 1);
+
+    newpath = PyUnicode_FromWideChar(s, wcslen(s));
+    if (newpath == NULL)
+        goto exit;
+
+    result = newpath;
+
+exit:
+    PyMem_Free(s);
+    return result;
+}
+
 /* Get the code object associated with the module specified by
    'fullname'. */
 static PyObject *
@@ -1465,11 +1717,26 @@ get_module_code(ZipImporter *self, PyObject *fullname,
     PyObject *code = NULL, *toc_entry, *subname;
     PyObject *path, *fullpath = NULL;
     struct st_zip_searchorder *zso;
+    time_t mtime = 0;
 
     subname = get_subname(fullname);
     if (subname == NULL)
         return NULL;
 
+    toc_entry = get_code_from_pycache(self, subname, &mtime);
+    if (toc_entry != NULL) {
+        code = get_code_from_data(self, 0, 1, mtime, toc_entry);
+        if (code != Py_None) {
+            if (code != NULL && p_modpath != NULL)
+                *p_modpath = strip_pycache_from_path(PyTuple_GetItem(toc_entry, 0));
+            if (p_ispackage != NULL)
+                *p_ispackage = 0;
+            Py_DECREF(subname);
+            return code;
+        }
+        Py_DECREF(code);
+    }
+
     path = make_filename(self->prefix, subname);
     Py_DECREF(subname);
     if (path == NULL)
@@ -1487,7 +1754,6 @@ get_module_code(ZipImporter *self, PyObject *fullname,
                                self->archive, (int)SEP, fullpath);
         toc_entry = PyDict_GetItem(self->files, fullpath);
         if (toc_entry != NULL) {
-            time_t mtime = 0;
             int ispackage = zso->type & IS_PACKAGE;
             int isbytecode = zso->type & IS_BYTECODE;
 
@@ -1500,6 +1766,17 @@ get_module_code(ZipImporter *self, PyObject *fullname,
             Py_CLEAR(fullpath);
             if (p_ispackage != NULL)
                 *p_ispackage = ispackage;
+
+            if (zso->type & IS_SOLIB) {
+                if (p_modpath != NULL) {
+                    *p_modpath = PyTuple_GetItem(toc_entry, 0);
+                    Py_INCREF(*p_modpath);
+                }
+                code = Py_None;
+                Py_INCREF(code);
+                goto exit;
+            }
+
             code = get_code_from_data(self, ispackage,
                                       isbytecode, mtime,
                                       toc_entry);
@@ -1509,10 +1786,8 @@ get_module_code(ZipImporter *self, PyObject *fullname,
                 Py_DECREF(code);
                 continue;
             }
-            if (code != NULL && p_modpath != NULL) {
-                *p_modpath = PyTuple_GetItem(toc_entry, 0);
-                Py_INCREF(*p_modpath);
-            }
+            if (code != NULL && p_modpath != NULL)
+                *p_modpath = strip_pycache_from_path(PyTuple_GetItem(toc_entry, 0));
             goto exit;
         }
         else
@@ -1554,6 +1829,51 @@ static struct PyModuleDef zipimportmodule = {
     NULL
 };
 
+static Py_ssize_t get_dynload_file_tab_len(void)
+{
+    int i = 0;
+    while (_PyImport_DynLoadFiletab[i])
+        i++;
+    return i;
+}
+
+static int init_search_order(void)
+{
+    struct st_zip_searchorder *zso;
+    const char **suffix;
+    int i;
+
+    snprintf(zip_searchorder_base[0].suffix, sizeof(zip_searchorder_base[0].suffix),
+             "%c__pycache__%c__init__.%s.pyc", SEP, SEP, _PySys_ImplCacheTag);
+    snprintf(zip_searchorder_base[1].suffix, sizeof(zip_searchorder_base[0].suffix),
+             "%c__init__.py", SEP);
+
+    zip_searchorder = PyMem_Malloc(sizeof(struct st_zip_searchorder) *
+                                   (zip_searchorder_base_len + get_dynload_file_tab_len() + 1));
+    if (zip_searchorder == NULL) {
+        PyErr_SetString(PyExc_MemoryError, "zipimport: out of memory");
+        return -1;
+    }
+
+    for (i = 0, zso = zip_searchorder_base; i < zip_searchorder_base_len; zso++, i++)
+        zip_searchorder[i] = *zso;
+
+    for (suffix = _PyImport_DynLoadFiletab; *suffix != NULL; suffix++) {
+        assert(strlen(*suffix) < sizeof(zso->suffix));
+        zso = &zip_searchorder[i];
+        memset(zso->suffix, 0, sizeof(zso->suffix));
+        strncpy(zso->suffix, *suffix, sizeof(zso->suffix) - 1);
+        zso->type = IS_SOLIB;
+        i++;
+    }
+
+    /* sentinel */
+    zip_searchorder[i].suffix[0] = '\0';
+    zip_searchorder[i].type = 0;
+
+    return 0;
+}
+
 PyMODINIT_FUNC
 PyInit_zipimport(void)
 {
@@ -1562,9 +1882,8 @@ PyInit_zipimport(void)
     if (PyType_Ready(&ZipImporter_Type) < 0)
         return NULL;
 
-    /* Correct directory separator */
-    zip_searchorder[0].suffix[0] = SEP;
-    zip_searchorder[1].suffix[0] = SEP;
+    if (init_search_order() < 0)
+        return NULL;
 
     mod = PyModule_Create(&zipimportmodule);
     if (mod == NULL)
